apps/prairielearn/src/components/DeleteCourseInstanceModal.tsx

```diff
@@ -0,0 +1,194 @@
+import { useEffect, useRef, useState } from 'preact/hooks';
+
+import type { CourseInstance } from '../lib/db-types.js';
+
+export function DeleteCourseInstanceModal({
+  shortName,
+  enrolledCount,
+  csrfToken,
+}: {
+  shortName: CourseInstance['short_name'];
+  enrolledCount: number;
+  csrfToken: string;
+}) {
+  const [step, setStep] = useState(1);
+  const [checks, setChecks] = useState({
+    impact: false,
+    irreversible: false,
+    recovery: false,
+  });
+  const [confirmationText, setConfirmationText] = useState('');
+
+  const allChecked = Object.values(checks).every(Boolean);
+  const canContinue = step === 1 ? allChecked : confirmationText === shortName;
+
+  const labelId = 'deleteCourseInstanceModalLabel';
+
+  // Reset state whenever the Bootstrap modal fully hides (after fade animation)
+  const modalRef = useRef<HTMLDivElement | null>(null);
+  useEffect(() => {
+    const el = modalRef.current;
+    if (!el) return;
+
+    const handleHidden = () => {
+      setStep(1);
+      setChecks({ impact: false, recovery: false, irreversible: false });
+      setConfirmationText('');
+    };
+
+    // Bootstrap dispatches 'hidden.bs.modal' after the hide transition completes
+    el.addEventListener('hidden.bs.modal', handleHidden);
+    return () => {
+      el.removeEventListener('hidden.bs.modal', handleHidden);
+    };
+  }, []);
+
+  // Focus the confirmation text input when moving to the second step.
+  const confirmationTextRef = useRef<HTMLInputElement | null>(null);
+  useEffect(() => {
+    // eslint-disable-next-line react-you-might-not-need-an-effect/no-event-handler
```

Comment:
A comment explaining that this is done in an effect so it occurs on next render would be nice. (or maybe the fact I recognize that now is enough...)

Reply:
I'll let you raise that if you want to.

Reply:
There is an `autoFocus` attribute we could use -- `autoFocus={step === 2}`